File: wrappers/java/phe/src/test/java/virgil/crypto/phe/PheClientTest.java
package virgil.crypto.phe;

import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import java.nio.charset.StandardCharsets;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

public class PheClientTest {

	private PheClient client;
	private PheServer server;

	@Before
	public void setup() {
		this.client = new PheClient();
		this.client.setupDefaults();
		assertTrue("C context should be set", this.client.cCtx > 0);

		this.server = new PheServer();
		this.server.setupDefaults();
		assertTrue("C context should be set", this.server.cCtx > 0);
	}

	@After
	public void teardown() {
		this.client.close();
		this.server.close();
	}

	@Test
	public void testFullFlowRandomCorrectPwdShouldSucceed() throws PheException {
		byte[] password = "password".getBytes(StandardCharsets.UTF_8);

		PheServerGenerateServerKeyPairResult serverKeyPair = server.generateServerKeyPair();
		assertNotNull(serverKeyPair);

		byte[] serverPrivateKey = serverKeyPair.serverPrivateKey;
		assertNotNull(serverPrivateKey);
		assertEquals(32, serverPrivateKey.length);

		byte[] serverPublicKey = serverKeyPair.serverPublicKey;
		assertNotNull(serverPublicKey);
		assertEquals(65, serverPublicKey.length);

		byte[] clientPrivateKey = this.client.generateClientPrivateKey(); // {privateKey}
		assertNotNull(clientPrivateKey);

		this.client.setKeys(clientPrivateKey, serverPublicKey); // void

		byte[] serverEnrollment = this.server.getEnrollment(serverPrivateKey, serverPublicKey);
		assertNotNull(serverEnrollment);
		assertTrue(serverEnrollment.length > 0);

		PheClientEnrollAccountResult clientEnrollAccount = this.client.enrollAccount(serverEnrollment, password);
		assertNotNull(clientEnrollAccount);

		byte[] clientEnrollmentRecord = clientEnrollAccount.enrollmentRecord;
		assertNotNull(clientEnrollmentRecord);

		byte[] clientAccountKey = clientEnrollAccount.accountKey;
		assertNotNull(clientAccountKey);
		assertEquals(32, clientAccountKey.length);

		byte[] clientCreateVerifyPasswordRequest = this.client.createVerifyPasswordRequest(password,
				clientEnrollmentRecord);
		assertNotNull(clientCreateVerifyPasswordRequest);
		assertTrue(clientCreateVerifyPasswordRequest.length > 0);

		byte[] serverVerifyPassword = this.server.verifyPassword(serverPrivateKey, serverPublicKey,
				clientCreateVerifyPasswordRequest);
		assertNotNull(serverVerifyPassword);

		byte[] clientCheckResponseAndDecrypt = this.client.checkResponseAndDecrypt(password, clientEnrollmentRecord,
				serverVerifyPassword);
		assertNotNull(clientCheckResponseAndDecrypt);
		assertEquals(32, clientCheckResponseAndDecrypt.length);
		assertArrayEquals(clientAccountKey, clientCheckResponseAndDecrypt);
	}

	@Test
	public void testRotationRandomRotationServerPublicKeysMatch() throws PheException {
		PheServerGenerateServerKeyPairResult serverKeyPair = this.server.generateServerKeyPair();
		assertNotNull(serverKeyPair);

		byte[] serverPrivateKey = serverKeyPair.serverPrivateKey;
		assertNotNull(serverPrivateKey);

		byte[] serverPublicKey = serverKeyPair.serverPublicKey;
		assertNotNull(serverPublicKey);

		PheServerRotateKeysResult serverRotateKeys = this.server.rotateKeys(serverPrivateKey);
		assertNotNull(serverRotateKeys);

		byte[] serverRotatedPrivateKey = serverRotateKeys.newServerPrivateKey;
		assertNotNull(serverRotatedPrivateKey);
		assertEquals(32, serverRotatedPrivateKey.length);

		byte[] serverRotatedPublicKey = serverRotateKeys.newServerPublicKey;
		assertNotNull(serverRotatedPublicKey);
		assertEquals(65, serverRotatedPublicKey.length);

		byte[] serverUpdateToken = serverRotateKeys.updateToken;
		assertNotNull(serverUpdateToken);
		assertTrue(serverUpdateToken.length > 0);

		byte[] clientPrivateKey = this.client.generateClientPrivateKey();
		assertNotNull(clientPrivateKey);
		assertEquals(32, clientPrivateKey.length);

		this.client.setKeys(clientPrivateKey, serverRotatedPublicKey);

		PheClientRotateKeysResult clientRotateKeys = this.client.rotateKeys(serverUpdateToken);
		assertNotNull(clientRotateKeys);

		byte[] clientNewPrivateKey = clientRotateKeys.newClientPrivateKey;
		assertNotNull(clientNewPrivateKey);
		assertEquals(32, clientNewPrivateKey.length);

		byte[] serverNewPublicKey = clientRotateKeys.newServerPublicKey;
		assertNotNull(serverNewPublicKey);
		assertEquals(65, serverNewPublicKey.length);

		assertEquals(serverPublicKey.length, serverNewPublicKey.length);
		assertEquals(clientPrivateKey.length, clientNewPrivateKey.length);
	}

}

File: wrappers/wasm/tests/phe/PheClient.test.js
const initPhe = require('../../dist/phe/node.cjs');
const { hexToUint8Array } = require('../utils');

describe('PheClient', () => {
  let phe;
  let pheClient;

  beforeEach(async () => {
    phe = await initPhe();
    pheClient = new phe.PheClient();
    pheServer = new phe.PheServer();
    pheClient.setupDefaults();
    pheServer.setupDefaults();
  });

  describe('enrollAccount', () => {
    it('should work', () => {
      const serverKeyPair = pheServer.generateServerKeyPair();
      const clientKeyPair = pheServer.generateServerKeyPair();
      pheClient.setKeys(clientKeyPair.serverPrivateKey, serverKeyPair.serverPublicKey);
      const enrollment = pheServer.getEnrollment(
        serverKeyPair.serverPrivateKey,
        serverKeyPair.serverPublicKey,
      );
      const password = hexToUint8Array('70617373776f7264');
      const { enrollmentRecord, accountKey } = pheClient.enrollAccount(enrollment, password);
      expect(enrollmentRecord).toBeInstanceOf(Uint8Array);
      expect(accountKey).toBeInstanceOf(Uint8Array);
    });
  });

  describe('createVerifyPasswordRequest', () => {
    it('should work', () => {
      const serverKeyPair = pheServer.generateServerKeyPair();
      const clientKeyPair = pheServer.generateServerKeyPair();
      pheClient.setKeys(clientKeyPair.serverPrivateKey, serverKeyPair.serverPublicKey);
      const enrollment = pheServer.getEnrollment(
        serverKeyPair.serverPrivateKey,
        serverKeyPair.serverPublicKey,
      );
      const password = hexToUint8Array('70617373776f7264');
      const { enrollmentRecord, accountKey } = pheClient.enrollAccount(enrollment, password);
      const request = pheClient.createVerifyPasswordRequest(password, enrollmentRecord);
      expect(request).toBeInstanceOf(Uint8Array);
    });
  });

  describe('checkResponseAndDecrypt', () => {
    it('should work', () => {
      const serverKeyPair = pheServer.generateServerKeyPair();
      const clientKeyPair = pheServer.generateServerKeyPair();
      pheClient.setKeys(clientKeyPair.serverPrivateKey, serverKeyPair.serverPublicKey);
      const enrollment = pheServer.getEnrollment(
        serverKeyPair.serverPrivateKey,
        serverKeyPair.serverPublicKey,
      );
      const password = hexToUint8Array('70617373776f7264');
      const { enrollmentRecord, accountKey } = pheClient.enrollAccount(enrollment, password);
      const request = pheClient.createVerifyPasswordRequest(password, enrollmentRecord);
      const response = pheServer.verifyPassword(
        serverKeyPair.serverPrivateKey,
        serverKeyPair.serverPublicKey,
        request,
      );
      const myAccountKey = pheClient.checkResponseAndDecrypt(password, enrollmentRecord, response);
      expect(myAccountKey.toString()).toBe(accountKey.toString());
    });
  });
});
